### ISSUE TITLE
X509_NAME_set error handling is incorrect

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#4324]
```
int X509_NAME_set(X509_NAME **xn, X509_NAME *name)
{
    X509_NAME *in;

    if (!xn || !name)
        return (0);

    if (*xn != name) {
        in = X509_NAME_dup(name);
        if (in != NULL) {
            X509_NAME_free(*xn);
            *xn = in;
        }
    }
    return (*xn != NULL);
}
```

If X509_NAME_dup fails (e.g., due to malloc failure), the local "in" will be set to NULL. This causes the subsequent condition to fail, and thus prevents the input "xn" from being set (actually a good thing, in and of itself). The input "xn" has already been checked against NULL previously, and has not been written to since. So, we know the final line will return a non-zero value (ie, success) despite not actually having set the input to the new value.

I don't see any particular security implications for this behavior, but certainly the whole "silent failure" business could create difficult debugging situations for application developers.

**Comment by richsalz** [comment]
What about this:
```
int X509_NAME_set(X509_NAME **xn, X509_NAME *name)
{   
    if ((name = X509_NAME_dup(name) == NULL)
        return 0;
    X509_NAME_Free(*xn);
    *xn = in;
    return 1;
}   
```

**Comment by jwalch** [comment]
I'm pretty sure that's not what you meant. It's not syntactically correct, and you'd potentially be leaking memory.

I don't think there's a way to avoid the local temporary here. Ideally, the function shouldn't modify either input in the error case. I don't see a way to achieve those semantics without stashing the X509_NAME_dup() result in a local temporary before making assignment decisions.

**Comment by richsalz** [comment]
Yeah, there's a typo.  But nothing is modified if the dup fails.

**Comment by levitte** [comment]
> I don't see a way to achieve those semantics without stashing the X509_NAME_dup() result in a local temporary before making assignment decisions.

`name` *is* a temporary storage in that function.  Whatever's assigned to it doesn't affect anything for the caller...

**Comment by jwalch** [comment]
Good point; pass by value means the caller retains the original pointer. I misinterpreted the intent behind the existing semantics (NULL check on `name` before assignment to local temporary).

**Comment by richsalz** [comment]
See #4326

**Comment by richsalz** [comment]
Merged to master, 1.0.2, 1.1.0  Closing the issue.
